synapse/handlers/pagination.py

```diff
@@ -72,13 +72,138 @@ def __init__(self, hs):
         self.auth = hs.get_auth()
         self.store = hs.get_datastore()
         self.clock = hs.get_clock()
+        self.config = hs.config
 
         self.pagination_lock = ReadWriteLock()
         self._purges_in_progress_by_room = set()
         # map from purge id to PurgeStatus
         self._purges_by_id = {}
         self._event_serializer = hs.get_event_client_serializer()
 
+        if self.config.retention_enabled:
+            # Run the purge jobs described in the configuration file.
+            for job in self.config.retention_purge_jobs:
+                self.clock.looping_call(
+                    self.purge_history_for_rooms_in_range,
+                    job["interval"],
+                    job["range_min"], job["range_max"],
+                )
+
+    @defer.inlineCallbacks
+    def is_event_outdated(self, event_ts, room_id):
+        """Checks whether the event with a given timestamp should be considered outdated
+        according to the room's or server's retention policy.
+
+        Args:
+            event_ts (int): Timestamp of the event in milliseconds.
+            room_id (str): Room in which the event was sent.
+
+        Returns:
+            bool: True if the event should be considered as outdated, False otherwise.
+        """
+        if self.config.retention_enabled:
+            try:
+                retention_policy = yield self.store.get_retention_policy_for_room(
+                    room_id
+                )
+
+                max_lifetime = retention_policy.get("max_lifetime")
+                if max_lifetime is None:
+                    max_lifetime = self.config.retention_max_lifetime
+
+                oldest_allowed_ts = self.clock.time_msec() - max_lifetime
+
+                if oldest_allowed_ts > event_ts:
+                    # Purge jobs are likely to get out of sync between servers, so we
+                    # need to ignore events that we have likely already purged.
+                    defer.returnValue(True)
+            except StoreError as e:
+                # Consider the event as not outdated if we can't verify whether we should
+                # ignore it.
+                logger.warning(
+                    "Error when looking up the retention policy for room %s: %s" % (
+                        room_id, e,
+                    )
+                )
+                defer.returnValue(False)
+
+        defer.returnValue(False)
+
+    @defer.inlineCallbacks
+    def purge_history_for_rooms_in_range(self, min_ts, max_ts):
+        """Purge outdated events from every room which maximum retention period is within
+        ]min_ts ; max_ts]. If the configured maximum period in the server's configuration
```

Comment:
> (it would be good to indicate here if the limits are inclusive or not)

That's what I meant by `]min_ts ; max_ts]`. I'll spell it out rather than use the mathematical representation of the range.

Reply:
ohh, sorry. I'm not familiar with that notation. I'd write this `(min_ts, max_ts]` following https://en.wikipedia.org/wiki/Interval_(mathematics)#Notations_for_intervals, but spelling it out is probably better.

Reply:
Right, in my case I've been so used to seeing and using the `]min_ts ; max_ts]` notation that I kinda forgot that the other existed. So yeah, spelling it out is definitely better.